rules/template-indent.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const stripIndent = require('strip-indent');
+const {replaceTemplateElement} = require('./fix/index.js');
+
+const MESSAGE_ID_IMPROPERLY_INDENTED_TEMPLATE = 'template-indent';
+const messages = {
+	[MESSAGE_ID_IMPROPERLY_INDENTED_TEMPLATE]: 'Templates should be properly indented. Selector: {{ selector }}',
+};
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const options = {
+		tags: ['outdent', 'dedent', 'gql', 'sql', 'html', 'styled'],
+		functions: ['dedent', 'stripIndent'],
+		selectors: [],
+		comments: ['HTML', 'indent'],
+		...context.options[0],
+	};
+
+	options.comments = options.comments.map(comment => comment.toLowerCase());
+
+	const selectors = [
+		...options.tags.map(tag => `TaggedTemplateExpression[tag.name="${tag}"] > .quasi`),
+		...options.functions.map(fn => `CallExpression[callee.name="${fn}"] > .arguments:first-child`),
+		...options.selectors,
+	];
+
+	/** @param {string} selector */
+	const getTemplateLiteralHandler = selector =>
+		/** @param {import('@babel/core').types.TemplateLiteral} node */
+		node => {
+			if (node.type !== 'TemplateLiteral') {
+				return;
+			}
+
+			const delimiter = '__PLACEHOLDER__' + Math.random();
+			const joined = node.quasis.map(q => q.value.raw).join(delimiter);
+
+			if (!joined.includes('\n')) {
+				return;
+			}
+
+			const dedented = stripIndent(joined).trim();
+
+			const startLine = context.getSourceCode().lines[node.loc.start.line - 1];
+			const marginMatch = startLine.match(/^(\s*)\S/);
+			const parentMargin = marginMatch ? marginMatch[1] : '';
+			const tabs = parentMargin.startsWith('\t');
+			const indent = tabs ? '\t' : '  ';
+			const templateMargin = parentMargin + indent;
+
+			const fixed = '\n'
+				+ dedented
+					.split('\n')
+					.map(line => line && line !== '\r' ? templateMargin + line : line)
+					.join('\n')
+					.trimEnd()
+				+ '\n'
+				+ parentMargin;
+
+			if (fixed === joined) {
+				return;
+			}
+
+			context.report({
+				node,
+				messageId: MESSAGE_ID_IMPROPERLY_INDENTED_TEMPLATE,
+				data: {
+					selector,
+				},
+				fix: fixer => fixed
+					.split(delimiter)
+					.map((replacement, index) => replaceTemplateElement(fixer, node.quasis[index], replacement)),
+			});
+		};
+
+	return {
+		...Object.fromEntries(selectors.map(selector => [selector, getTemplateLiteralHandler(selector)])),
+		/**
+		 * @param {import('@babel/core').types.TemplateLiteral} node
+		 */
+		TemplateLiteral: node => {
```

Comment:
If user have `TemplateLiteral` selector, it should report on any template, but I think current implementation won't, because it's overridden.

Is it possible to add only one `TemplateLiteral` listener, and use `esquery.matches` to test node with  `options.selectors`?

Reply:
I reduced down to one listener for functions/tags/custom selectors, and one for comments (where the comments one is not added if `options.comments` is empty). The single listener for functions/tags/custom selectors uses `const {matches} = require('./selectors/index.js');` to generate one big selector matcher string.